### ISSUE TITLE
Encoding API is difficult to use

### ISSUE THREAD

**Issue by bos** [tibbe/cassava#60]
I've been trying to use cassava for what seem like some simple purposes, and have so far gotten nowhere.

My use case is inside criterion, where I want to write out the header of a CSV file in one part of the program, and then (somewhere not closely related) write out individual lines/records to the file.

From what I can figure out, to support this I need to write a `ToNamedHeader` instance, then create a `Header` value, then `encodeByNameWith` with two custom `EncodeOptions` values, one that makes sure a header is written out (for my first use case), and another that does not (for my second).

It honestly took me a couple of hours of banging my head against the API docs to even figure this out, because it's not easy to see how the various types and functions inter-relate. The code that I think I would need to write to actually do all of this feels like a lot of work, so I haven't switched over.

What I really hoped I was signing up for was a much more lightweight API:

* write a `Generic` instance for my type, then have the boilerplate `ToNamedHeader` instance generated for me with one line of code: `instance ToNamedRecord MyType`

* an `encodeHeader` function

* an `encodeRow` function

* variants of `encodeHeader` and `encodeRow` that could accept a `Header` spec to control ordering, but also variants that don't need this repetition if you don't care

**Comment by tibbe** [comment]
I've worked on this during ZuriHac. I have a prototype here: https://github.com/tibbe/cassava/tree/builder

Note that if you're using the builder (or any builder, like the bytestring one) in I/O it cannot be streaming, as it's only streaming when used in a lazy foldr-style computation, which I/O isn't. You can still use it in an accumulator style computation for convenience, but you don't get the streaming.

**Comment by tibbe** [comment]
This also contains all the Generics machinery. Example:

```haskell
{-# LANGUAGE DeriveGeneric #-}

data Person = Person { name :: !Text , salary :: !Int }
    deriving Generic

instance FromRecord Person
instance ToRecord Person
```

Usage:

```
>>> encode [Person ("John" :: Text) 27]
"John,27\r\n"
```

This also in other places these classes are used, as in the new builder module.

**Comment by tibbe** [comment]
Try out the newly released 0.4.3.0.
